### PR TITLE
*: support new placement rules syntax

### PR DESCRIPTION
**Pull request by xhebox**
### What problem does this PR solve? 

The syntax part for pingcap/tidb#26581. One extra change: `constraints` is moved from `unreserved` to `notkeywords`, since mysql does not have such keyword.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported variable/fields change

Related changes

 - Need to be included in the release note

### CODE REVIEW

File: parser.y [hunk 1]
```diff
@@ -1562,13 +1622,13 @@ PlacementOptions:
...
+	"ADD" "PLACEMENT" "POLICY" OldPlacementOptions
 	{
 		spec := $4.(*ast.PlacementSpec)
 		spec.Tp = ast.PlacementAdd
 		$$ = spec
 	}
-|	"ALTER" "PLACEMENT" "POLICY" PlacementOptions
+|	"ALTER" "PLACEMENT" "POLICY" OldPlacementOptions
 	{
 		spec := $4.(*ast.PlacementSpec)
```
Comment (morgo) [inline parser.y:1633]: Should we create an issue to track removing the `OldPlacementOptions`? I assume it will be a good first issue.

Comment (xhebox) [reply parser.y:1633]: I've added it in the todo of pingap/tidb#26581 :)

There will be a conflict between the implementation. Once the new syntax is implemented, the old syntax must be removed.

File: parser.y [hunk 2]
```diff
@@ -691,6 +690,16 @@ import (
 	learner               "LEARNER"
 	verboseType           "VERBOSE"
 	voter                 "VOTER"
+	constraints           "CONSTRAINTS"
```
Comment (AilinKid) [inline parser.y:693]: we used to put this alphabetically

File: ast/ddl.go [hunk 3]
```diff
@@ -71,12 +71,25 @@ const (
 	DatabaseOptionCharset
 	DatabaseOptionCollate
 	DatabaseOptionEncryption
+	DatabaseOptionPlacementPrimaryRegion
+	DatabaseOptionPlacementRegions
+	DatabaseOptionPlacementFollowerCount
+	DatabaseOptionPlacementVoterCount
+	DatabaseOptionPlacementLearnerCount
+	DatabaseOptionPlacementSchedule
+	DatabaseOptionPlacementConstraints
```
Comment (AilinKid) [inline ast/ddl.go:80]: offset calculation base on the consecutive definition here. 
once another definition inserted among here, a reflecting map is necessary.

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

File: ast/ddl.go (hunk 3 above)
Comment (xhebox) [reply ast/ddl.go:80]: It is intended. All options are same for db/table/partition anyway. I could assign every table/db enum one specific number like `0x3000 + iota`.

File: parser.y (hunk 2 above)
Comment (xhebox) [reply parser.y]: Done

**Comment by AilinKid** [comment]
Screenshot_20210816_172122

Kenny has some suggestions with my draft PR (in fact), PTAL

**Comment by xhebox** [comment]
> 
> 
> Screenshot_20210816_172122
> 
> Kenny has some suggestions with my draft PR (in fact on your base), PTAL

It is addressed.

**Review by AilinKid — APPROVED** [review]

**Comment by AilinKid** [comment]
sure, better `make parser` again to resolve the conflicts

**Review by morgo — APPROVED** [review]

**Pull request merged**